src/Traits/HasRoles.php

```diff
@@ -107,19 +119,30 @@ public function assignRole(...$roles)
             ->each(function ($role) {
                 $this->ensureModelSharesGuard($role);
             })
-            ->map->id
-            ->all();
+            ->map(function ($role) {
+                return ['id' => $role->id, 'values' => PermissionRegistrar::$teams && !is_a($this, Permission::class) ?
+                    [PermissionRegistrar::$teamsKey => app(PermissionRegistrar::class)->getPermissionsTeamId()] : []
+                ];
+            })
+            ->pluck('values', 'id')->toArray();
 
         $model = $this->getModel();
 
         if ($model->exists) {
-            $this->roles()->sync($roles, false);
+            if (PermissionRegistrar::$teams && !is_a($this, Permission::class)) {
+                $this->roles()->wherePivot(PermissionRegistrar::$teamsKey, app(PermissionRegistrar::class)->getPermissionsTeamId())->sync($roles, false);
+            } else {
+                $this->roles()->sync($roles, false);
+            }
             $model->load('roles');
         } else {
             $class = \get_class($model);
 
             $class::saved(
                 function ($object) use ($roles, $model) {
+                    if ($model->getKey() != $object->getKey()) {
+                        return;
+                    }
```

Comment:
Thanks for your detailed description of this bug in the original post. 
Is it possible to move this bug-fix to its own separate PR (both for HasRoles and HasPermissions)? I'd like to merge this fix separately from this Teams feature.

Also, is there another test that should be added to prevent this bug? A few different approaches have been taken to get to this point, including model/object comparisons, and I'd like to make sure we don't re-break things again. Refs: #975 and #1663

Reply:
>model/object comparisons 

That is unnecessary beacuse, `$class::saved` comes from `$class=\get_class($model);`, so model and object always have the same class. i test on mi two aplications and it works, I hope I don't break something, because I recently started checking this code